src/DominantJuice.sol

```diff
@@ -0,0 +1,396 @@
+    /// @return weight The weight that project tokens should get minted relative to. This is useful for optionally customizing how many tokens are issued per payment.
+    /// @return memo A memo to be forwarded to the event. Useful for describing any new actions that are being taken.
+    /// @return delegateAllocations Amount to be sent to delegates instead of adding to local balance. Useful for auto-routing funds from a treasury as payment come in.
+    function payParams(JBPayParamsData calldata _data)
+        external
+        virtual
+        override
+        returns (uint256 weight, string memory memo, JBPayDelegateAllocation3_1_1[] memory delegateAllocations)
+    {
+        // The if statements are the only changes from the payParams() template.
+        // This is also not a payable function, so calls with msg.value will revert.
+        if (getBalance() == 0) revert DataSourceNotInitialized();
+        if (_data.amount.value < minimumPledgeAmount) revert AmountIsBelowMinimumPledge(minimumPledgeAmount);
+
+        // Forward the default weight received from the protocol.
+        weight = _data.weight;
+        // Forward the default memo received from the payer.
+        memo = _data.memo;
+        // Add `this` contract as a Pay Delegate so that it receives a `didPay` call. Don't send any funds to the delegate (keep all funds in the treasury).
+        delegateAllocations = new JBPayDelegateAllocation3_1_1[](1);
+        delegateAllocations[0] = JBPayDelegateAllocation3_1_1(this, 0, "");
+
+        return (weight, memo, delegateAllocations);
+    }
+
+    /// @notice Received hook from the payment terminal after a payment.
+    /// @dev Reverts if the calling contract is not one of the project's terminals.
+    /// @dev This example implementation reverts if the payer isn't on the allow list.
+    /// @param _data Standard Juicebox project payment data.
+    /// See https://docs.juicebox.money/dev/api/data-structures/jbdidpaydata3_1_1/.
+    function didPay(JBDidPayData3_1_1 calldata _data) external payable virtual override {
+        // Make sure the caller is a terminal of the project, and that the call is
+        // being made on behalf of an interaction with the correct project.
+        if (
+            msg.value != 0 || !directory.isTerminalOf(projectId, IJBPaymentTerminal(msg.sender))
+                || _data.projectId != projectId
+        ) revert InvalidPaymentEvent(msg.sender, _data.projectId, msg.value);
+
+        if (block.timestamp >= cycleExpiryDate) revert CycleHasExpired();
+
+        // Get payer address and amount paid from JB DidPay data struct
+        address payer = _data.payer;
+        JBTokenAmount memory amount = _data.amount;
+        uint256 paymentAmount = amount.value;
+
+        // Update storage variables
+        totalAmountPledged += paymentAmount;
+        pledgedAmount[payer] = paymentAmount;
+        pledgers.push(payer);
+        orderToPledgerToAmount[pledgeOrder][payer] = paymentAmount;
```

Comment:
this is odd to me. Why is this a mapping to a mapping? the second mapping always has a single value in it because the pledgeOrder is always increased after this call, so there is exactly one payer.

Reply:
When the individual refund bonus is calculated, it was a structure I created to deal with a pledger pledging twice, so that a pledger would have two distinct payments that would be weighted differently in the calculation. (In order to simplify, I asked Scott if we could just require pledgers that wanted to pledge twice to just use a different wallet address. And he seemed open to the idea, but wanted me to try this first to see if it worked.) I saw you had some more comments/questions on the formula, so I'll answer more there.

Reply:
Can you link to what lines assume # of pledgers = # of pledges?

The reason for the weird structure was to make the normalizing order and weights math work. 

I like the list of structs. Would that play nice with the OZ library you recommended?

Reply:
lists of structs don't support the set concept because the struct isn't hashable. but again, not sure you need the set anyway because your calcs all just assume you have to loop over the whole list, so you get no efficiency improvements.

Not sure you even need all the individual orders anyway. If you redo the math, then you could just keep sums for everything all along.

- keep a total per pledger (not really needed, but might be nice to track)
- total amount
- keep a weighted total per pledger (based on when their order arrived)
- keep a weighted total amount (based on when the order arrived)

Then the math per pledger is really easy: total amount * pledgerWeightedTotal / weighted total (rounded down)